### ISSUE TITLE
Affinity session azure front door and app service Unable to unprotect the message.State

### ISSUE THREAD

**Issue by haris-begluk** [AzureAD/microsoft-identity-web#2599]
Failure on azure entra callback. 

I have two environments: dev and test. In the dev environment, I'm using only app service to deploy Duende Identity (2 instances) along with Entra authentication as an external provider.
In the test environment, I'm using Azure Front Door in front of the App Service where Duende Identity (2 instances) is deployed.
In the dev environment, I enabled session affinity because I was encountering the error:
'Unable to unprotect the message.State,' and that fixed the issue, making it work fine.

However, in the test environment. I'm still facing the same problem, even with session affinity enabled on the App Service.
I'm pretty sure that something is not configured properly with my Azure Front Door.
The issue occurs randomly on the callback (/.auth/login/aad/callback) when I try to log in with Entra ID. 

I tried to bypass the Azure Front Door in the test environment, and that did the trick—no failures. 
This kind of proves that I am missing something in Azure Front Door. 

How to host duende server behind the AFD when having external callbacks?
Does the session affinity cookie guarantee success?
Do I need to enable session affinity on the Azure Front Door itself? 
What do I need to do to make it work? 

By the way I have data protection enabled and DP keys are persisted in the database. 

I would highly appreciate any guidance.  
Thanks in advance.

```
An error was encountered while handling the remote login.
Exception
System.Exception: An error was encountered while handling the remote login.
 ---> System.Exception: Unable to unprotect the message.State.
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
   at Duende.IdentityServer.Hosting.FederatedSignOut.AuthenticationRequestHandlerWrapper.HandleRequestAsync() in /_/src/IdentityServer/Hosting/FederatedSignOut/AuthenticationRequestHandlerWrapper.cs:line 52
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Duende.IdentityServer.Hosting.DynamicProviders.DynamicSchemeAuthenticationMiddleware.Invoke(HttpContext context) in /_/src/IdentityServer/Hosting/DynamicProviders/DynamicSchemes/DynamicSchemeAuthenticationMiddleware.cs:line 48
   at Duende.IdentityServer.Hosting.BaseUrlMiddleware.Invoke(HttpContext context) in /_/src/IdentityServer/Hosting/BaseUrlMiddleware.cs:line 28
```

**Comment by jmprieur** [comment]
@haris-begluk you might want to ask the question in the asp.net core repo (https://github.com/dotnet/aspnetcore), or in the front door support. I don't think that this is related to Microsoft.Identity.Web.

**Comment by haris-begluk** [comment]
I apologize for creating an issue here. I now realize I posted in the wrong place, was misled by the error message: 'System.Exception: Unable to unprotect the message.State,' which I found had been discussed previously here.

Anyway for the benefit of the community, I will share my experience.
I'm glad to say that I managed to solve the problem by using distributed caching as described in the documentation. 
It appears that an affinity session was not a good option in my case, and I was unable to make it work with Azure Front Door. Additionally, requests were slower.

The issue stemmed from my identity server being configured to cache state for OIDC handlers in the distributed cache, like this: services.AddOidcStateDataFormatterCache(). However, I hadn't configured any real distributed cache, which meant that every instance held state values in memory. This would result in a failed callback if the entrance callback ended up in the wrong instance.

In short, as soon as I added SQL Server distributed caching, OIDC states were cached in my case in SQL Server, and I encountered no more issues, eliminating the need to use affinity sessions.
